src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ExhaustivenessComputer.java

```diff
@@ -113,55 +136,85 @@ public boolean exhausts(JCExpression selector, List<JCCase> cases) {
-            return true;
+            return ExhaustivenessResult.ofExhaustive();
         }
 
         for (Entry<Symbol, Set<Symbol>> e : enum2Constants.entrySet()) {
             if (e.getValue().isEmpty()) {
                 patternSet.add(new BindingPattern(e.getKey().type));
             }
         }
-        Set<PatternDescription> patterns = patternSet;
-        Set<Set<PatternDescription>> seenFallback = new HashSet<>();
-        boolean useHashes = true;
         try {
-            boolean repeat = true;
-            while (repeat) {
-                Set<PatternDescription> updatedPatterns;
-                updatedPatterns = reduceBindingPatterns(selector.type, patterns);
-                updatedPatterns = reduceNestedPatterns(updatedPatterns, useHashes);
-                updatedPatterns = reduceRecordPatterns(updatedPatterns);
-                updatedPatterns = removeCoveredRecordPatterns(updatedPatterns);
-                repeat = !updatedPatterns.equals(patterns);
-                if (checkCovered(selector.type, patterns)) {
-                    return true;
-                }
-                if (!repeat) {
-                    //there may be situation like:
-                    //class B permits S1, S2
-                    //patterns: R(S1, B), R(S2, S2)
-                    //this might be joined to R(B, S2), as B could be rewritten to S2
-                    //but hashing in reduceNestedPatterns will not allow that
-                    //disable the use of hashing, and use subtyping in
-                    //reduceNestedPatterns to handle situations like this:
-                    repeat = useHashes && seenFallback.add(updatedPatterns);
-                    useHashes = false;
-                } else {
-                    //if a reduction happened, make sure hashing in reduceNestedPatterns
-                    //is enabled, as the hashing speeds up the process significantly:
-                    useHashes = true;
-                }
-                patterns = updatedPatterns;
+            CoverageResult coveredResult = computeCoverage(selector.type, patternSet, PatternEquivalence.STRICT);
+            if (coveredResult.covered()) {
+                return ExhaustivenessResult.ofExhaustive();
             }
-            return checkCovered(selector.type, patterns);
+
+            Set<String> details =
+                    this.computeMissingPatternDescriptions(selector.type, coveredResult.incompletePatterns())
+                        .stream()
+                        .flatMap(pd -> {
+                            if (pd instanceof BindingPattern bp && enum2Constants.containsKey(bp.type.tsym)) {
+                                Symbol enumType = bp.type.tsym;
+                                return enum2Constants.get(enumType).stream().map(c -> enumType.toString() + "." + c.name);
+                            } else {
+                                return Stream.of(pd.toString());
```

Comment:
As discussed offline, eager string generation will render the diagnostic arguments completely opaque to the formatter. This would mean that no `where` clauses will be generated, and unambiguous qualifiers will not be omitted.

Reply:
Also, don't forget `JCDiagnostic.MultilineDiagnostic`, which we use in overloads to render repetitive fragments in tabular format.